sys/build.rs

```diff
@@ -1,31 +1,207 @@
+                version = Some(parse_geos_version(&raw_version.to_string_lossy()));
+            }
+            None => {
+                panic!("GEOS_VERSION must be set");
+            }
+        }
+    } else {
+        // try to detect using pkg-config, if available
+        version = detect_geos_via_pkg_config();
+
+        // fall back to try using geos-config
+        if version.is_none() {
+            version = detect_geos_via_geos_config();
+        }
+
+        if version.is_none() {
+            panic!("Could not detect GEOS using pkg-config or geos-config");
+        }
+    }
+
+    let version = version.unwrap();
+
+    let min_geos_version = Version::parse(MINIMUM_GEOS_VERSION).unwrap();
+    if version < min_geos_version {
+        panic!(
+            "GEOS version {}.{}.{} is older than the minimum supported version {}.{}.{}",
+            version.major,
+            version.minor,
+            version.patch,
+            min_geos_version.major,
+            min_geos_version.minor,
+            min_geos_version.patch
+        );
     }
 
-    // TODO: handle library versions like this!
-    //
-    // pkg_config::probe_library("geos_c")
-    //     .map(|lib| {
-    //         if lib.version.starts_with("2.") {
-    //             println!(r#"cargo:rustc-cfg=feature="v2""#);
-    //         }
-    //     })
-    //     .expect("GEOS library not found");
+    // resolve user-requested version (via specific version feature, e.g., "v3_10")
+    // to the correct pre-built binding; their available GEOS must be >= requested
+    // pre-built binding version
+
+    let mut binding_version = Version::parse(MINIMUM_GEOS_VERSION).unwrap();
+
+    if cfg!(feature = "v3_7_0") {
+        binding_version = Version::new(3, 7, 0);
+    }
+
+    if cfg!(feature = "v3_8_0") {
+        binding_version = Version::new(3, 8, 0);
+    }
+
+    if cfg!(feature = "v3_9_0") {
+        binding_version = Version::new(3, 9, 0);
+    }
+
+    if cfg!(feature = "v3_10_0") {
+        binding_version = Version::new(3, 10, 0);
+    }
+
+    if version < binding_version {
+        panic!("You requested a version of GEOS ({}.{}) that is greater than your installed GEOS version ({}.{}.{})", binding_version.major, binding_version.minor, version.major, version.minor, version.patch);
+    }
+
+    // copy requested prebuilt binding (if exists) to output directory
+    let binding_path = PathBuf::from(format!(
+        "prebuilt-bindings/geos_{}.{}.rs",
+        binding_version.major, binding_version.minor
+    ));
+
+    // this shouldn't happen except when a new version feature is added but the
+    // binding has not yet been created
+    if !binding_path.exists() {
+        panic!("No pre-built bindings available for requested GEOS version {}.{}\nUse features to select an available version.", binding_version.major, binding_version.minor);
+    }
+
+    std::fs::copy(&binding_path, &out_path).expect("Can't copy bindings to output directory");
```

Comment:
That is definitely not great (especially for releases). Please generate a cfg instead (and then link to the correct one).

Reply:
You actually already generate a `cfg` just below. Then in `sys/lib.rs`, you would have:

```rust
#[cfg(feature = "geos_sys_x_y_z_0")]
include!("the_bindgen_file.rs");
#[cfg(feature = "geos_sys_x2_y2_z2_0")]
include!("another_bindgen_file.rs");
// ...
```

Reply:
Ok, I think I understand - using the `cfg` in `lib.rs` lets us linked to a binding file as is without copying / renaming it - so it sidesteps the need for line 201; we'd still keep the version verification logic in `build.rs`.  This does mean that `lib.rs` will then also need to be updated for each new version feature, in addition to `build.rs`.

Reply:
This is unfortunately proving more difficult than I expected, though this is a bit over my head.

if we use 

```rust
#[cfg(feature = "geos_sys_3_10_0")]
include!("../prebuilt-bindings/geos_3.10.rs");
```

it never includes the bindings, even though we set `cargo:rustc-cfg=geos_sys_3_10_0` at the end of `build.rs`.  (maybe I'm misunderstanding how that is set)

If we instead use the version features as they are defined in `Cargo.toml`

```rust 
#[cfg(feature = "v3_10_0")]
include!("../prebuilt-bindings/geos_3.10.rs");
```

we run into collisions because of the way dependencies are setup between the version features (which is what we want for the main crate).  Enabling v3_10_0 pulls in the bindings for v3_9_0, v3_8_0, etc. which then causes compilation errors.

The way `build.rs` was originally constructed, whichever version feature was the highest number had the highest precedence, and the logic ensured that only one binding file would ever be bundled in - which ultimately is what we want if we go the `cfg` route; I'm just not seeing how to achieve that given my limited understanding of `cfg`.

Reply:
Nevermind, my confusion was how to refer to the `cfg` value; they aren't features when set from `build.rs`.

We needed this instead, then things work properly (except for `dox` feature, still sorting that out):

```rust
#[cfg(geos_sys_3_10)]
include!("../prebuilt-bindings/geos_3.10.rs");
```

Reply:
You can use the `feature` too. For example like this:

```rust
#[cfg(all(feature = "v3_10_0", not(feature = "v3_11_0"))]
include!("../prebuilt-bindings/geos_3.10.rs");
```

Every time you have to check if the "upper" version feature isn't set. This might be a bit more straightforward since it relates to the cargo feature directly and not the one generated by cargo. Which makes me think: if it is only used to know which file to include, maybe we can stop generating this `cfg`?

Reply:
Excellent suggestion; this got around the issue I was having with testing lower versions.